---
title: "RamseyR"
author: "Gabriel Balsa Cantú"
organization: "UNAM - LIIGH"
date: "Tuesday, August 18, 2015"
output: html_document
---
#Introduction
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

The following implementation is an open source reconstruction of the published algorithm by Stephen A. Ramsey, Bioinformatics Vol. 26, 2010 <http://bioinformatics.oxfordjournals.org/>.

The following code is open source, **built under R and C++ using the standard RStudio toolset including Rtools** in a Windows 7 64 bit mainframe.
The used pacakges were:

- ggplot2
- Rcpp
- dplyr
- reshape2

Additional functions are provided via gitHub distribution, and should be runned "as is". <https://github.com/GbalsaC/TFBSValley>

## Purpose & Aims
Development of a new prediction method of potential **transcriptional factors binding sites (TFBS)** combining ChIP-Seq  data obtained from mammal cell regulatory regions. 
This method's prediction rate should outperform the actual ChIP-Seq based techniques. 


### Initial Configurations and Setup
The test case includes **aa.csv** which is data obtained from the original algorithm simulating signal.

> When used on other local machines, **setwd()** should be changed to the cloned or forked repository accordingly.
**signal** is imported (or should be) as a numeric vector, as follows:

```{r}
setwd("B:/Documents/R/TFBSValley")
signal<-as.numeric(read.csv("aa.csv", header=F)[1,])
### Requires multiplot.R, Rcpp, ggplot
head(signal);

```

# Code Body
### Definition and Parameters

The first process is defining the function, which requires two inputs:

1. **signal**: A numeric vector, on the first commits of this code it's ideally a csv row vector. On the following commits, the ideal file should be a BamFile, commonly used in Bioinformatics.
2. **index**: The number of the first element within a genomic sequence or a chromosome range sequence. It's purpose is simply identifying the sequence's domain, therefore it's an optional input.

Additionally, the function requires loading the necessary libraries.

```{r, eval=FALSE}
RamseyR <- function(signal, index=1){
        ##Packages, sourceCpp has rebuild False
        library(ggplot2);
        library("dplyr");
        library(Rcpp);
        library("reshape2");
        sourceCpp("LemireR.cpp");
        source('multiplot.R');


```

```{r, echo=FALSE, warning=FALSE}
        index=1;
        library(knitr)
        library(ggplot2);
        library("dplyr");
        library(Rcpp);
        library("reshape2");
        source('~/R/functions/multiplot.R');
```


## C++ Code Engine lemireR wrapper
The main filter engine is built in C++ for performance, using the **Rcpp** package it can be called as a native R function **lemireR(a=Array, w=Window, offset=, Depth=(0<1) )**.

This engine is software independent and can be used as a standalone command by replacing vectors for double* -> new double appropiately as C++ native array indexes.

```{r engine='Rcpp'}
#include <Rcpp.h>
#include <iostream>
#include <vector>
#include <deque>
#include <fstream>      // fstream
#include <string>
#include <sstream>
#include <algorithm>    // copy
#include <math.h>
#include <iterator>     // ostream_operator
using namespace Rcpp;
using namespace std;
typedef unsigned int uint;

//Allocate memory for size, window and pointer for dynamic memory allocation
double *maxval;
double *minval;

long sizeArray;
long index, index2;
//std::vector<double> a;
std::deque<int> U, L;
int n;


// [[Rcpp::export]]
NumericMatrix lemireR(NumericVector a, int w, int offset, double Depth){
        index = 0; index2 = 0;
        sizeArray = a.size() + 1;
        //A = new double[sizeArray];
        NumericVector A1(sizeArray);
        NumericVector A2(sizeArray);        
	NumericMatrix A(a.size(), 5);
	minval = new double[sizeArray];
	maxval = new double[sizeArray];



	for (long i = 1; i < a.size(); ++i)
	{
		if (i >= w)
		{
			maxval[i - w] = a[U.size()>0 ? U.front() : i - 1];
			minval[i - w] = a[L.size()>0 ? L.front() : i - 1];
		}
		if (a[i]>a[i - 1])
		{
			L.push_back(i - 1);
			if (i == w + L.front())
			{
				L.pop_front();
			}
			while (U.size()>0)
			{
				if (a[i] <= a[U.back()])
				{
					if (i == w + U.front())
					{
						U.pop_front();
					}
					break;
				}
				U.pop_back();
			}
			//End While
		}
		else
		{
			U.push_back(i - 1);
			if (i == w + U.front())
			{
				U.pop_front();
			}
			while (L.size()>0)
			{
				if (a[i] >= a[L.back()])
				{
					if (i == w + L.front())
					{
						L.pop_front();
					}
					break;
				}//END IF
				L.pop_back();
			}//END While
		}//End If-Else
	}//END FOR
	maxval[a.size() - w] = a[U.size()>0 ? U.front() : a.size() - 1];
	minval[a.size() - w] = a[L.size()>0 ? L.front() : a.size() - 1];

	//myfile is optionally only usable as csv output
	for (long l = 1; l < sizeArray; l++){
		if (l < sizeArray - (w))
		{
        		A(l-1+(w/2), 0) = maxval[l];
                        //A(0, l-1) = maxval[l];
        		//myfile << maxval[l] << ",";
		}
                else if(l < sizeArray - (w/2))
		{
			A(l-1+(w/2), 0) = 0;
			//myfile << 0 << ",";
		}
                /* 
                */
	}
	//myfile << "\n";

        n = w + offset;
        // C++ code for V(pos) in Ramsey's Matlab 
	for (long pos = 1; pos < sizeArray-1; pos++){
                if ((pos >= n) && (pos < (sizeArray - w))){
			A(pos, 1) = min(A(lrint(pos + offset + w / 2), 0), A(lrint(pos - offset - w / 2),0));
                        
		}
		else {
			A(pos, 1) = 0;
		}
                //Create logical comparison of A(pos,1) with a[]
                //Consider previous pos if, if 0 then it's false for all comparisons
                // Of the type A(pos,1)*Depth > a[pos]
                if((A(pos, 1)*Depth)>a[pos]){
                        A(pos, 2)=A(pos, 1);
                }
                
                
	}
        

	//myfile.close();
	delete[] maxval;
	delete[] minval;
	return A;
}
	

```


## Preparing the function
Similarly to the original algorithm, default parameters are defined as follows:
```{r, echo=TRUE}
        ## Default Values and Parameters
        Window=100;
        Offset = 10;
        Depth = 0.9;
        sigma=6;

```

### Reducing noise
The next step involves noise reduction by:

1. Using a normal distribution to get the range for a convolution function
2. Using convolution to smooth signal according to a predefined normal distribution

```{r, echo=TRUE}
        sigma2=exp((sigma-1)*log(2)/2);
        X=qnorm(1e-3,0,sigma2);
        window2=dnorm(round(X):-round(X),0,sigma2);
        wSize=length(window2);
        ones=vector(mode="double",length=wSize);
        for(i in 1:length(ones)){ones[i]<-1;}
        vector=ones/wSize;
        
        ##Cloning variables for modification individually 
        ##is memory intensive, and should not duplicate memory usage
        M = convolve(vector, rev(as.numeric(signal)), type="open")[ceiling(wSize/2):ceiling(length(signal)-1+(wSize/2))];
        newWin = Window-Offset;
        

```

### Calling C++ function.
The function lemireR is built with Rcpp package and RStudio as mentioned in the introduction, the code can [be accessed here](/blob/master/LemireR.cpp). 
The code returns a five column matrix to which the index numeration is binded as reference.

The columns are ordered as:

1. **Index**: a localization numeral indicating the specified region on a genetic element.

2. **MD**: The Max filter signal, adjusted for domain +/- newWin, in order to obtain half a window after and before the original signal.

3. **Vpos**: The minimum value between two elements, the comparison of minimum elements is made by **+/- offset + (newWin/2)**, this is the valley's minimum value, used to determine valley regions by comparison.

4. **Vlog**: Using Vpos x Depth, using a boolean comparison, set to true wherever (Vpos x Depth) > signal.

```{r}
        
        M = lemireR(a=M, w=newWin+1, offset=10, Depth=0.8); 
        
        length <- dim(M)[1];
        M<-cbind(c(index:(index-1+length)),M);
        colnames(M)<-c('Index','MD', 'Vpos', 'Vlog', 'Begin', 'End');
        head(M[264:274,]);
```

In retrospect, the valley regions are actually comparison between the signal and the smallest value of max filter's +/- newWindow region.

```{r, echo=FALSE}
        M2<-melt(as.data.frame(M), id="Index", measure=c("MD", "Vpos", "Vlog"));
        ggplot(data=M2, aes(x=Index, y=value, colour=variable))+geom_line();
        ggsave(file="ramseyR.png");

```

## Adding Begin and End Regions of the matrix
Using nested for loops, we can add the new columns wich give the regions of acceptable valleys from begining and end index.

```{r}
        count=0;
        start=0;
        for(i in 1:length){
                if(M[i,'Vlog']>0)
                {
                        if(start==0){start=M[i,'Index'];}
                        M[i,'Begin']=start;
                        count=count+1;        
                        
                }else
                {
                        if(start != 0)
                        {
                                ### Then it's the first time entering 0
                                #### All elements from start to actual of the End column
                                #### should save this index value
                                M[start:(M[i,'Index']-1),'End']=M[i,'Index'];
                                start=0;
                                count=0;
                        }
                        ##Do nothing else
                }

        }
         head(M[264:274,]);

```

Finally, converting the output as data.frame for as factor functionality.
```{r}
        M<-as.data.frame(M);
        M
```

